### PR TITLE
Fix healing whole file bitrot

### LINKED ISSUE #7121: Bitrot Detection/Healing Fails when using Highwayhash256 after recent changes
If I set the default bitrot detection algorithm to highwayhash256 (the legacy, not the current 256S), I receive "unexpected EOF" errors when running healing, including on the hidden .config buckets.

Version: 74c2048ea9ad0982123623d5e0a184f42490762b

## Expected Behavior
Expected behavior is with a support bitrot detection algorithm, the mc admin heal command should run to completion, and there should be no errors reported on any servers.

## Current Behavior

Instead, when I run mc admin heal --recursive myminio/buck using highwayhash256 as the default algorithm, I receive sometimes on the client (e.g. insufficient space for rebuild) and, and consistently on the target.

## Possible Solution
Unknown.

## Steps to Reproduce (for bugs)
1. Set the default bitrot algorithm on each server to HighwayHash256 (not HighwayHash256S), can be hard coded in cmd/bitrot.go
2.   Start a cluster of 2 servers, each with 2 disks
3.  Upload some data
4. Run mc admin heal on the bucket where data was uploaded

```
ifadams@anchor:~/minio_d1/buck/5MB$ mc admin heal --recursive myminio
 ◓  buck/5MB
    0/1 objects; 0 B in 1s
    ┌────────┬───┬─────────────────────┐
    │ Green  │ 3 │  75.0% █████████    │
    │ Yellow │ 0 │   0.0%              │
    │ Red    │ 0 │   0.0%              │
    │ Grey   │ 1 │  25.0% ███          │
    └────────┴───┴─────────────────────┘

Server Side Log:
API: Heal(bucket=buck, object=5MB)
Time: 14:14:42 PST 01/18/2019
RemoteHost: 192.168.1.160
Error: unexpected EOF
       1: cmd/xl-v1-healing-common.go:188:cmd.disksWithAllParts()
       2: cmd/xl-v1-healing.go:278:cmd.healObject()
       3: cmd/xl-v1-healing.go:671:cmd.xlObjects.HealObject()
       4: cmd/xl-sets.go:1259:cmd.(*xlSets).HealObject()
       5: cmd/admin-heal-ops.go:778:cmd.(*healSequence).healObject()
       6: cmd/admin-heal-ops.go:750:cmd.(*healSequence).healBucket.func1()
       7: pkg/sync/errgroup/errgroup.go:55:errgroup.(*Group).Go.func1()

<occurs 3 more times>
```
And here is an example xl.json

```
{"version":"1.0.1","format":"xl","stat":{"size":5242880,"modTime":"2019-01-18T21:55:21.32338001Z"},"erasure":{"algorithm":"klauspost/reedsolomon/vandermonde","data":2,"parity":2,"blockSize":10485760,"index":3,"distribution":[3,4,1,2],"checksum":[{"name":"part.1","algorithm":"highwayhash256","hash":"a46aaa059334c0e00dc82c6d9ab995daa2d1bac1a987fd7b8db26758d874875e"}]},"minio":{"release":"DEVELOPMENT.GOGET"},"meta":{"content-type":"application/octet-stream","etag":"0d2d1bff0e7921709424c6920b3c605a"},"parts":[{"number":1,"name":"part.1","etag":"","size":5242880,"actualSize":5242880}]}i
```

## Context
Impacting R&D that uses Minio as a target system

## Your Environment

DEVELOPMENT.GOGET
Ubuntu 16.04

### PR DESCRIPTION
**Pull request by krishnasrinivas**
Fixes https://github.com/minio/minio/issues/7121

## Description

Healing code had these issues:
* bitrotCheckFile was passing bigger buffer than size of the file causing errors
* incorrect checksum were passed to newBitrotReader() causing it to fail
* newXLMetaFromXLMeta() was not copying xlMeta.Stat

## Regression

Yes PR #7004

## How Has This Been Tested?

By simulating various heal scenarios:
* removing a file from a disk
* causing bitrot corruption

Both wholefile bitrot and streaming bitrot were verified.

## Types of changes

- [ ] Bug fix (non-breaking change which fixes an issue)
- [ ] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to change)

## Checklist:

- [ ] My code follows the code style of this project.
- [ ] My change requires a change to the documentation.
- [ ] I have updated the documentation accordingly.
- [ ] I have added unit tests to cover my changes.
- [ ] I have added/updated functional tests in [mint](https://github.com/minio/mint). (If yes, add `mint` PR # here: )
- [ ] All new and existing tests passed.

### CODE REVIEW

**Comment by krishnasrinivas** [comment]
Requesting reviews from @harshavardhana @vadmeste

**Comment by minio-ops** [comment]
## Mint Automation
| Test | Result |
|:--- |:--- |
| mint-tls.sh | :heavy_check_mark: |
| mint-compression-xl.sh | :heavy_check_mark: |
| mint-xl.sh | :heavy_check_mark: |
| mint-gateway-nas.sh | :heavy_check_mark: |
| mint-large-bucket.sh | :heavy_check_mark: |
| mint-compression-fs.sh | :heavy_check_mark: |
| mint-worm.sh | :heavy_check_mark: |
| mint-fs.sh | :heavy_check_mark: |
| mint-dist-xl.sh | :heavy_check_mark: |

**Comment by vadmeste** [comment]
```diff
diff --git a/cmd/bitrot_test.go b/cmd/bitrot_test.go
index 94b32594..36a6ec7e 100644
--- a/cmd/bitrot_test.go
+++ b/cmd/bitrot_test.go
@@ -24,7 +24,7 @@ import (
 	"testing"
 )
 
-func TestBitrotReaderWriter(t *testing.T) {
+func testBitrotReaderWriterAlgo(t *testing.T, bitrotAlgo BitrotAlgorithm) {
 	tmpDir, err := ioutil.TempDir("", "")
 	if err != nil {
 		log.Fatal(err)
@@ -41,7 +41,7 @@ func TestBitrotReaderWriter(t *testing.T) {
 
 	disk.MakeVol(volume)
 
-	writer := newBitrotWriter(disk, volume, filePath, 35, HighwayHash256S, 10)
+	writer := newBitrotWriter(disk, volume, filePath, 35, bitrotAlgo, 10)
 
 	_, err = writer.Write([]byte("aaaaaaaaaa"))
 	if err != nil {
@@ -61,7 +61,7 @@ func TestBitrotReaderWriter(t *testing.T) {
 	}
 	writer.(io.Closer).Close()
 
-	reader := newStreamingBitrotReader(disk, volume, filePath, 35, HighwayHash256S, 10)
+	reader := newStreamingBitrotReader(disk, volume, filePath, 35, bitrotAlgo, 10)
 	b := make([]byte, 10)
 	if _, err = reader.ReadAt(b, 0); err != nil {
 		log.Fatal(err)
@@ -76,3 +76,9 @@ func TestBitrotReaderWriter(t *testing.T) {
 		log.Fatal(err)
 	}
 }
+
+func TestAllBitrotAlgorithms(t *testing.T) {
+	for bitrotAlgo, _ := range bitrotAlgorithms {
+		testBitrotReaderWriterAlgo(t, bitrotAlgo)
+	}
+}
```

This is a quick diff of bitrot_test.go to check all bitrot algorithms (default and no default algorithms)

**Comment by harshavardhana** [comment]
Thanks @vadmeste

**Comment by vadmeste** [comment]
@krishnasrinivas healing a highwayhash256 object still throws an error, also above unit tests throw an EOF error.

**Comment by krishnasrinivas** [comment]
> @krishnasrinivas healing a highwayhash256 object still throws an error

how to reproduce this @vadmeste 

> also above unit tests throw an EOF error.

you have to do this and it will work:
```
-       reader := newStreamingBitrotReader(disk, volume, filePath, 35, HighwayHash256S, 10)
+       reader := newBitrotReader(disk, volume, filePath, 35, bitrotAlgo, bitrotWriterSum(writer), 10)
```

File: cmd/xl-v1-object_test.go [hunk 1]
```diff
@@ -320,6 +321,8 @@ func TestHealing(t *testing.T) {
 
 	// After heal the meta file should be as expected.
 	if !reflect.DeepEqual(xlMetaPreHeal, xlMetaPostHeal) {
+		fmt.Println(xlMetaPreHeal)
```
Comment (harshavardhana) [inline cmd/xl-v1-object_test.go]: This needs to be removed

**Review by harshavardhana — CHANGES_REQUESTED** [review]

File: cmd/xl-v1-object_test.go (hunk 1 above)
Comment (krishnasrinivas) [reply cmd/xl-v1-object_test.go]: :+1:

**Review by harshavardhana — DISMISSED** [review]

**Comment by minio-ops** [comment]
## Mint Automation
| Test | Result |
|:--- |:--- |
| mint-tls.sh | :heavy_check_mark: |
| mint-compression-xl.sh | :heavy_check_mark: |
| mint-xl.sh | :heavy_check_mark: |
| mint-gateway-nas.sh | :heavy_check_mark: |
| mint-large-bucket.sh | :heavy_check_mark: |
| mint-compression-fs.sh | :heavy_check_mark: |
| mint-worm.sh | :heavy_check_mark: |
| mint-fs.sh | :heavy_check_mark: |
| mint-dist-xl.sh | :heavy_check_mark: |

**Comment by vadmeste** [comment]
LGTM but Travis is failing @krishnasrinivas

**Review by vadmeste — APPROVED** [review]
LGTM and tested

**Comment by minio-ops** [comment]
## Mint Automation
| Test | Result |
|:--- |:--- |
| mint-tls.sh | :heavy_check_mark: |
| mint-compression-xl.sh | :heavy_check_mark: |
| mint-xl.sh | :heavy_check_mark: |
| mint-gateway-nas.sh | :heavy_check_mark: |
| mint-large-bucket.sh | :heavy_check_mark: |
| mint-compression-fs.sh | :heavy_check_mark: |
| mint-worm.sh | :heavy_check_mark: |
| mint-fs.sh | :heavy_check_mark: |
| mint-dist-xl.sh | :heavy_check_mark: |

**Review by harshavardhana — APPROVED** [review]

**Pull request merged**